NDTensors/src/dense/dense.jl

```diff
@@ -3,104 +3,135 @@
 #
 using LinearAlgebra: BlasFloat
 
-struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
-  data::VecT
-  function Dense{ElT,VecT}(data::AbstractArray) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+struct Dense{ElT,DataT<:AbstractArray} <: TensorStorage{ElT}
+  data::DataT
+  function Dense{ElT,DataT}(data::DataT) where {ElT,DataT<:AbstractVector}
+    @assert ElT == eltype(DataT)
+    return new{ElT,DataT}(data)
   end
 
-  # Special case of Vector
-  function Dense{ElT,VecT}(data::Vector) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
-  end
-
-  # Special case of Array which gets turned into a Vector
-  function Dense{ElT,VecT}(data::Array) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(vec(data))
+  function Dense{ElT,DataT}(data::DataT) where {ElT,DataT<:AbstractArray}
+    println("Only Vector-based datatypes are currently supported.")
+    throw(TypeError)
   end
 end
 
-# TODO: make these more general, move to tensorstorage.jl
-datatype(::Type{<:Dense{<:Any,DataT}}) where {DataT} = DataT
-
-function set_datatype(storagetype::Type{<:Dense}, datatype::Type{<:AbstractVector})
-  return Dense{eltype(datatype),datatype}
+#Start with high information constructors and move to low information constructors
+function Dense{ElT,DataT}() where {ElT,DataT<:AbstractArray}
+  return Dense{ElT,DataT}(DataT())
 end
 
-function set_datatype(storagetype::Type{<:Dense}, datatype::Type{<:AbstractArray})
-  return error(
-    "Setting the `datatype` of the storage type `$storagetype` to a $(ndims(datatype))-dimsional array of type `$datatype` is not currently supported, use an `AbstractVector` instead.",
-  )
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,DataT}(inds::Tuple) where {ElT,DataT<:AbstractArray}
+  return Dense{ElT,DataT}(generic_zeros(DataT, dim(inds)))
 end
 
-# TODO: Define a generic `dense` for `Tensor`, `TensorStorage`.
-dense(storagetype::Type{<:Dense}) = storagetype
+function Dense{ElT,DataT}(dim::Integer) where {ElT,DataT<:AbstractArray}
+  return Dense{ElT,DataT}(generic_zeros(DataT, dim))
+end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+function Dense{ElT,DataT}(::UndefInitializer, inds::Tuple) where {ElT,DataT<:AbstractArray}
+  return Dense{ElT,DataT}(similar(DataT, dim(inds)))
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+function Dense{ElR,DataT}(data::AbstractArray) where {ElR,DataT<:AbstractArray}
+  data = convert(DataT, data)
+  return Dense{ElR,DataT}(data)
 end
 
+# This function is ill-defined. It cannot transform a complex type to real...
 function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
   return ElT == ElR ? Dense(data) : Dense(ElR.(data))
```

Comment:
Should be written in terms of `convert`?

Reply:
Oh yes, this accidentally got converted back to the old version which is what was on your branch